### PR TITLE
Mailbox 270: implement quota and update documentation

### PR DESCRIPTION
**Pull request by quynhn**
- For current limitation, the quota defines in java constant file

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java [hunk 1]
```diff
@@ -440,7 +441,7 @@
      * @param mailboxAnnotations    the list of annotation should be insert/udpate/delete
      * @throws MailboxException in case of selected mailbox does not exist
      */
-    void updateAnnotations(MailboxPath mailboxPath, MailboxSession session, List<MailboxAnnotation> mailboxAnnotations) throws MailboxException;
+    void updateAnnotations(MailboxPath mailboxPath, MailboxSession session, List<MailboxAnnotation> mailboxAnnotations) throws MailboxException, AnnotationException;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: Why adding this exception to signature when its already covered y MailboxException ?

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 2]
```diff
@@ -372,4 +374,32 @@ public void getAnnotationsByKeysWithAllDepthShouldRetriveAnnotationsWithAllDepth
...
+    @ContractTest
+    public void updateAnnotationsShouldThrowExceptionIfRequestCreateNewButMailboxIsOverLimit() throws MailboxException {
+        expected.expect(AnnotationException.class);
+        Assume.assumeTrue(mailboxManager.hasCapability(MailboxCapabilities.Annotation));
+        session = mailboxManager.createSystemSession(USER_2, LoggerFactory.getLogger("Test"));
+        MailboxPath inbox = MailboxPath.inbox(session);
+        mailboxManager.createMailbox(inbox, session);
+
+        ImmutableList.Builder<MailboxAnnotation> builder = ImmutableList.builder();
+        for (int i = 0; i < MailboxConstants.LIMIT_ANNOTATIONS_ON_MAILBOX; i++) {
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: You can avoid this for if you use a field parameter for this. You give the constant value by default, and use a setter to lower value with tests, so that you do not need flow control.

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 3]
```diff
@@ -372,4 +374,32 @@ public void getAnnotationsByKeysWithAllDepthShouldRetriveAnnotationsWithAllDepth
...
+            builder.add(MailboxAnnotation.newInstance(new MailboxAnnotationKey("/private/comment" + i), "AnyValue"));
+        }
+        mailboxManager.updateAnnotations(inbox, session, builder.build());
+
+        mailboxManager.updateAnnotations(inbox, session, ImmutableList.of(PRIVATE_ANNOTATION));
+    }
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: You should verify that : 
 - You do not throw when you have the exact count of annotations
 - You have the exact size of annotations

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 4]
```diff
@@ -372,4 +374,32 @@ public void getAnnotationsByKeysWithAllDepthShouldRetriveAnnotationsWithAllDepth
...
+    @ContractTest
+    public void updateAnnotationsShouldThrowExceptionIfAnnotationDataIsOverLimitation() throws MailboxException {
+        expected.expect(AnnotationException.class);
+        Assume.assumeTrue(mailboxManager.hasCapability(MailboxCapabilities.Annotation));
+        session = mailboxManager.createSystemSession(USER_2, LoggerFactory.getLogger("Test"));
+        MailboxPath inbox = MailboxPath.inbox(session);
+        mailboxManager.createMailbox(inbox, session);
+
+        mailboxManager.updateAnnotations(inbox, session, ImmutableList.of(MailboxAnnotation.newInstance(PRIVATE_KEY,
+            StringUtils.leftPad("AnyData", MailboxConstants.LIMIT_ANNOTATION_DATA + 1, '*'))));
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Not needed if the annotation size is customizable.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java [hunk 5]
```diff
@@ -98,6 +99,16 @@ public void insertAnnotation(MailboxId mailboxId, MailboxAnnotation mailboxAnnot
             .value(CassandraAnnotationTable.VALUE, mailboxAnnotation.getValue().get()));
     }
 
+    @Override
+    public boolean isExisted(MailboxId mailboxId, MailboxAnnotation mailboxAnnotation) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java]: The name is not understandable. Is it **exist** ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 6]
```diff
@@ -780,12 +781,23 @@ public void updateAnnotations(MailboxPath mailboxPath, MailboxSession session, L
         for (MailboxAnnotation annotation : mailboxAnnotations) {
             if (annotation.isNil()) {
                 annotationMapper.deleteAnnotation(mailboxId, annotation.getKey());
-            } else {
+            } else if (isValid(mailboxId, annotation, annotationMapper)) {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: You should allow annotations update if it decrease annotation size, even over the limit
Otherwise, if some obscur conflict occurs, one can not get under the quota, in some scenario.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 7]
```diff
@@ -780,12 +781,23 @@ public void updateAnnotations(MailboxPath mailboxPath, MailboxSession session, L
...
         }
     }
 
+    private boolean isValid(MailboxId mailboxId, MailboxAnnotation annotation, AnnotationMapper annotationMapper) throws AnnotationException {
+        if (annotation.size() > MailboxConstants.LIMIT_ANNOTATION_DATA) {
+            throw new AnnotationException("annotation too big.");
+        }
+        if (!annotationMapper.isExisted(mailboxId, annotation)
+            && annotationMapper.countAnnotations(mailboxId) >= MailboxConstants.LIMIT_ANNOTATIONS_ON_MAILBOX) {
+            throw new AnnotationException("too many annotations.");
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: Capital letter. Which is the allowed annotation  count ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java [hunk 8]
```diff
@@ -205,4 +205,26 @@ public void annotationShouldBeCaseInsentive() throws Exception {
...
+    }
+
+    @ContractTest
+    public void countAnnotationShouldReturnNumberOfAnnotationBelongToMailbox() throws Exception {
+        annotationMapper.insertAnnotation(mailboxId, PRIVATE_ANNOTATION);
+        annotationMapper.insertAnnotation(mailboxId, PRIVATE_ANNOTATION_UPDATE);
+        annotationMapper.insertAnnotation(mailboxId, SHARED_ANNOTATION);
+        annotationMapper.insertAnnotation(mailboxId, PRIVATE_USER_ANNOTATION);
+
+        assertThat(annotationMapper.countAnnotations(mailboxId)).isEqualTo(3);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java]: You added 4 annotations. Why is there just 3 remaining ?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java [hunk 9]
```diff
@@ -205,4 +205,26 @@ public void annotationShouldBeCaseInsentive() throws Exception {
...
+    }
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java]: You should try countAnnotations when there is no annotation (edge case)

File: server/src/site/xdoc/feature-protocols.xml [hunk 10]
```diff
@@ -78,6 +78,20 @@
         Note that MOVE extension (RFC-6851) needs a MailboxManager implementing the move function. For now, this is only done with
         the CassandraMailboxManager
       </p>
+      <p>
+        It supports the METADATA extension (RFC-5464) but
```
Comment (chibenwa) [inline server/src/site/xdoc/feature-protocols.xml]: :

File: server/src/site/xdoc/feature-protocols.xml [hunk 11]
```diff
@@ -78,6 +78,20 @@
...
+        <ul>
+          <li>
+            It does not support the server's metadata so far.
+          </li>
+          <li>
+            It does not support permission on metadata cause the ACL extension is not present.
```
Comment (chibenwa) [inline server/src/site/xdoc/feature-protocols.xml]: because

File: server/src/site/xdoc/feature-protocols.xml [hunk 12]
```diff
@@ -78,6 +78,20 @@
...
+          </li>
+          <li>
+            It's had the limitation values which is hardcode instead of configurable. Those are the number of annotations on mailbox (10) and the maximum size of an annotation (1024 bytes).
```
Comment (chibenwa) [inline server/src/site/xdoc/feature-protocols.xml]: First sentence is not grammatically correct... 
I would propose "The quota limitations on MailboxAnnotations are hardcoded. They concern the number of annotations per mailbox, as well as the size of an annotation (1024 bytes).

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 13]
```diff
@@ -780,12 +781,23 @@ public void updateAnnotations(MailboxPath mailboxPath, MailboxSession session, L
...
                 annotationMapper.deleteAnnotation(mailboxId, annotation.getKey());
-            } else {
+            } else if (isValid(mailboxId, annotation, annotationMapper)) {
                 annotationMapper.insertAnnotation(mailboxId, annotation);
             }
         }
     }
 
+    private boolean isValid(MailboxId mailboxId, MailboxAnnotation annotation, AnnotationMapper annotationMapper) throws AnnotationException {
+        if (annotation.size() > MailboxConstants.LIMIT_ANNOTATION_DATA) {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: Does annotation.size() verifies size of the name + value or just value ?

We should check RFC...

If it is not the case I can still abuse the implementation, and store arbitrary amount of data using annotations names.

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java (hunk 1 above)
Comment (quynhn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: It uses to custom the output message at MPT layer.

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java (hunk 2 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Done

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java (hunk 3 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: I tried and it took a lot of time but still don't know how to verify the called method so I extracted 2 functions testing instead of

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java (hunk 5 above)
Comment (quynhn) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java]: Done

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java (hunk 4 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Done

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 7 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: It's not capital letter cause it should be consistent output at MPT testing. No capital.
For the message, I followed the RFC at '4.3.  SETMETADATA Command'.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java (hunk 8 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java]: Cause the second one is just update

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java (hunk 9 above)
Comment (quynhn) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/AnnotationMapperTest.java]: Done

File: server/src/site/xdoc/feature-protocols.xml (hunk 10 above)
Comment (quynhn) [inline server/src/site/xdoc/feature-protocols.xml]: Done

File: server/src/site/xdoc/feature-protocols.xml (hunk 11 above)
Comment (quynhn) [inline server/src/site/xdoc/feature-protocols.xml]: Done

File: server/src/site/xdoc/feature-protocols.xml (hunk 12 above)
Comment (quynhn) [inline server/src/site/xdoc/feature-protocols.xml]: Done

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 13 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: RFC said "Annotation data size" so it's just value.
It said at: "4.1.  General Considerations"
"The server is allowed to impose limitations on the size of any one
   annotation or the total number of annotations for a single mailbox or
   for the server as a whole.  However, the server MUST accept an
   annotation data size of at least 1024 bytes, and an annotation count
   per server or mailbox of at least 10."

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java [hunk 14]
```diff
@@ -23,6 +23,7 @@
 import java.util.EnumSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: Why do you need this import?

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 15]
```diff
@@ -76,6 +78,8 @@
 
     @Rule
     public ExpectedException expected = ExpectedException.none();
+    @Rule
+    public ErrorCollector collector = new ErrorCollector();
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Where are you using your collector?

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 16]
```diff
@@ -372,4 +378,46 @@ public void getAnnotationsByKeysWithAllDepthShouldRetriveAnnotationsWithAllDepth
             .contains(PRIVATE_ANNOTATION, PRIVATE_CHILD_ANNOTATION, PRIVATE_GRANDCHILD_ANNOTATION);
     }
 
-}
+    @ContractTest
+    public void updateAnnotationsShouldThrowExceptionIfAnnotationDataIsOverLimitation() throws MailboxException {
+        expected.expect(AnnotationException.class);
+        Assume.assumeTrue(mailboxManager.hasCapability(MailboxCapabilities.Annotation));
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: I would like the Assume to be the first part of tests

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java [hunk 17]
```diff
@@ -372,4 +378,46 @@ public void getAnnotationsByKeysWithAllDepthShouldRetriveAnnotationsWithAllDepth
...
+        session = mailboxManager.createSystemSession(USER_2, LoggerFactory.getLogger("Test"));
+        MailboxPath inbox = MailboxPath.inbox(session);
+        mailboxManager.createMailbox(inbox, session);
+        mailboxManager.setLimitAnnotationSize(10);
+
+        mailboxManager.updateAnnotations(inbox, session, ImmutableList.of(MailboxAnnotation.newInstance(PRIVATE_KEY, "DataMoreThan10")));
+    }
+
+    @ContractTest
+    public void updateAnnotationsShouldBeOKIfRequestCreateNewAndMailboxIsNotOverLimit() throws MailboxException {
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: shouldUpdateAnnotationWhenRequestCreatesNewAndMailboxIsNotOverLimit

File: mailbox/api/src/test/java/org/apache/james/mailbox/mock/MockMailboxManager.java [hunk 18]
```diff
@@ -161,5 +161,6 @@ private void createMailbox(MailboxSession mailboxSession, MailboxPath mailboxPat
         }
         getMockMailboxManager().endProcessingRequest(mailboxSession);
     }
-
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/mock/MockMailboxManager.java]: noise

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java [hunk 19]
```diff
@@ -98,6 +99,16 @@ public void insertAnnotation(MailboxId mailboxId, MailboxAnnotation mailboxAnnot
             .value(CassandraAnnotationTable.VALUE, mailboxAnnotation.getValue().get()));
     }
 
+    @Override
+    public boolean exist(MailboxId mailboxId, MailboxAnnotation mailboxAnnotation) {
+        return !getAnnotationsByKeys(mailboxId, ImmutableSet.of(mailboxAnnotation.getKey())).isEmpty();
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java]: it looks very inefficient

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java [hunk 20]
```diff
@@ -98,6 +99,16 @@ public void insertAnnotation(MailboxId mailboxId, MailboxAnnotation mailboxAnnot
...
+    }
+
+    @Override
+    public int countAnnotations(MailboxId mailboxId) {
+        return getAllAnnotations(mailboxId).size();
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java]: and this too

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 21]
```diff
@@ -780,12 +783,23 @@ public void updateAnnotations(MailboxPath mailboxPath, MailboxSession session, L
...
             if (annotation.isNil()) {
                 annotationMapper.deleteAnnotation(mailboxId, annotation.getKey());
-            } else {
+            } else if (isValid(mailboxId, annotation, annotationMapper)) {
                 annotationMapper.insertAnnotation(mailboxId, annotation);
             }
         }
     }
 
+    private boolean isValid(MailboxId mailboxId, MailboxAnnotation annotation, AnnotationMapper annotationMapper) throws AnnotationException {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: I don't think it's a problem of being valid, it's more about being authorized to insert something new

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 22]
```diff
@@ -815,4 +829,14 @@ public boolean hasChildren(MailboxPath mailboxPath, MailboxSession session) thro
         Mailbox mailbox = mapper.findMailboxByPath(mailboxPath);
         return mapper.hasChildren(mailbox, session.getPathDelimiter());
     }
+
+    @Override
+    public void setLimitOfAnnotations(int limitOfAnnotations) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: setters, really ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java [hunk 23]
```diff
@@ -84,4 +84,20 @@
...
+     * @param mailboxId the selected mailbox
+     * @param mailboxAnnotation current annotation should be checked
+     * @return boolean should be 'true' if it already existed on store
+     */
+    boolean exist(MailboxId mailboxId, MailboxAnnotation mailboxAnnotation);
+
+    /**
+     * Getting total number of current annotation on mailbox
+     *
+     * @param mailboxId the selected mailbox
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java]: you can drop @param and  @result when they say the same thing as the method prototype + description

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java [hunk 24]
```diff
@@ -84,4 +84,20 @@
      */
     void insertAnnotation(MailboxId mailboxId, MailboxAnnotation mailboxAnnotation);
 
+    /**
+     * Checking the current annotation of selected mailbox is existed on store or not.
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java]: I guess you don't check the value but only if any entry has the same key already ?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java (hunk 14 above)
Comment (quynhn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MailboxManager.java]: Removed

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java (hunk 15 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: I tried to verify some functions should be called but I couldn't. Sorry that I forgot to remove it

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java (hunk 16 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Done

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java (hunk 17 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java]: Updated

File: mailbox/api/src/test/java/org/apache/james/mailbox/mock/MockMailboxManager.java (hunk 18 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/mock/MockMailboxManager.java]: Reset to the last one

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java (hunk 23 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java]: Done

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java (hunk 24 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/AnnotationMapper.java]: Updated the description

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 22 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: yes, it customs the value of limitation.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 21 above)
Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: it's not only insertion the new one but also for updating (in case of checking the size of annotation). I changed to "canInsertOrUpdate".

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java (hunk 20 above)
Comment (quynhn) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java]: Updated

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java (hunk 19 above)
Comment (quynhn) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java]: Updated

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java (hunk 22 above)
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: Is there any runtime command that allows to modify the limit ? If not, please pass by constructor.

Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: But there are so many parameters in constructor. Is there any convention for that? Eg: the number of function's parameters must less than 10
And one more thing, the StoreMailboxManager is main class so if we changed the constructor (with adding 2 new parameter just for annotation) will effect a lot, so many code should be changed. That seems not good I think. WDYT

Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: There's a lot of questions here.
First, about the number of parameters :
* Having too many parameters usually shows that the class does too much : that's clearly the case, but do you want to break it into small parts right now ?
* It's often painful because you don't want to call the constructor, here it's not a problem because of injection, but we would use a builder in another case

Second, the lifecycle of objects is really important : if you allow passing a configuration variable into a setter, does that mean that you can change at anytime ? Usually, we don't want to deal with this kind of complexity, and showing that the parameter is immutable and passed at construction time is so valuable that we prefer handling a refactoring than adding setters to a class.

To conclude about that : there are already setters in this class, so harm is already done. You can skip the refactoring about that but you should at least make it clear that this method should be called before `init()` and handle is value like other setters do.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxAnnotation.java [hunk 25]
```diff
@@ -24,7 +24,6 @@
 import com.google.common.base.Preconditions;
 
 public class MailboxAnnotation {
-
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxAnnotation.java]: Noise

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 26]
```diff
@@ -126,6 +127,9 @@
 
     private final MessageParser messageParser;
 
+    private int limitOfAnnotations;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: Not final ?

Comment (quynhn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: It cannot be final I think because it can be initial by 2 ways. One as default value and the other as custom value for testing.

Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: Make it final please, in both case I can now it at instanciation, and it will not change. (Optional constructor parameter might be the solution ?)

**Comment by chibenwa** [comment]
Rebase this branch on master ;-)

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by quynhn** [comment]
test this please
